workflow/artifacts/oss/oss.go

```diff
@@ -212,3 +223,136 @@ func isTransientOSSErr(err error) bool {
+		if err != nil {
+			return errors.InternalWrapError(err)
+		}
+		// build the name to be used in oss
+		nameInDir, err := filepath.Rel(dir, fpath)
+		if err != nil {
+			return errors.InternalWrapError(err)
+		}
+		fObjectName := filepath.Join(objectName, nameInDir)
+		//for local dir, create an oss dir explicitly;
+		//thus, for an empty local dir, an empty oss dir is also created.
+		if info.Mode().IsDir() {
+			// create oss dir
+			if !strings.HasSuffix(fObjectName, "/") {
+				fObjectName += "/"
+			}
+			err = bucket.PutObject(fObjectName, nil)
+			if err != nil {
+				return err
+			}
+		}
+		if !info.Mode().IsRegular() {
+			return nil
+		}
+
+		err = putFile(bucket, fObjectName, fpath)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+//IsOssErrCode tests if an err is an oss.ServiceError with the specified code
+func IsOssErrCode(err error, code string) bool {
+	if serr, ok := err.(oss.ServiceError); ok {
+		if serr.Code == code {
+			return true
+		}
+	}
+	return false
+}
+
+// IsDirectory tests if the key is acting like a oss directory. This just means it has at least one
+// object which is prefixed with the given key
+func IsOssDirectory(bucket *oss.Bucket, objectName string) (bool, error) {
+	if objectName == "" {
+		return true, nil
+	}
+	if !strings.HasSuffix(objectName, "/") {
+		objectName += "/"
+	}
+	rst, err := bucket.ListObjects(oss.Prefix(objectName), oss.MaxKeys(1))
+	if err != nil {
+		return false, err
+	}
+	if len(rst.CommonPrefixes)+len(rst.Objects) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// GetOssDirectory download a oss "directory" to local path
+func GetOssDirectory(bucket *oss.Bucket, objectName, path string) error {
+	files, err := ListOssDirectory(bucket, objectName)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		innerName, err := filepath.Rel(objectName, f)
+		if err != nil {
+			return fmt.Errorf("get Rel path from %s to %s error: %w", f, objectName, err)
+		}
+		fpath := filepath.Join(path, innerName)
+		if strings.HasSuffix(f, "/") {
+			err = os.MkdirAll(fpath, 0o700)
+			if err != nil {
+				return fmt.Errorf("mkdir %s error: %w", fpath, err)
+			}
+			continue
+		}
+		dirPath := filepath.Dir(fpath)
+		err = os.MkdirAll(dirPath, 0o700)
+		if err != nil {
+			return fmt.Errorf("mkdir %s error: %w", dirPath, err)
+		}
+
+		err = bucket.GetObjectToFile(f, fpath)
+		if err != nil {
+			log.Warnf("failed to load object %s to %s error: %v", f, fpath, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// ListOssDirectory lists all the files which are the descendants of the specified objectKey, if a file has suffix '/', then it is an oss directory
+func ListOssDirectory(bucket *oss.Bucket, objectKey string) (files []string, err error) {
+	if objectKey != "" {
+		if !strings.HasSuffix(objectKey, "/") {
+			objectKey += "/"
+		}
+	}
+
+	pre := oss.Prefix(objectKey)
+	const batchCount = 50
```

Comment:
Could you add a comment on why this is hard-coded

Reply:
this hard-coded "50" has no much meaning, it is just a number sounds reasonable pop up from my head when I was coding it. 50 is not too big that would cause timeout, and not two small that would cause too much network requests.
In fact, if `oss.MaxKeys` is omitted, 100 will be used internally by default.
Maybe I should not explicitly set this option here, I will remove it later.

Reply:
Yes let's remove it and use the default unless really necessary